### DISCUSSION TITLE
Callback FUNC_BUTTON_PRESSED fires without button action

### DISCUSSION THREAD

**Discussion by Heizelmann** [arendst/Tasmota#17830] [Q&A] [answered]
I try to use Callback _FUNC_BUTTON_PRESSED_ in my own driver code interface function and it triggers endlessly in quick succession without touching the push button.The button is connected to PIN 8 and configured as Button. The payload is correct and depends on the button press or release.

switch (function) {
    ...
    case FUNC_BUTTON_PRESSED:
          AddLog(LOG_LEVEL_INFO, PSTR("Button pressed:%d"),XdrvMailbox.payload);
          result = true;       
      break;      
  }

**Comment by barbudor** [comment, marked as answer]
It is my understanding that FUNC_BUTTON_PRESSED event is passed to all driver every <debounce_time> providing the current state of the button
So if the default debounce time of 50ms, that would be every 50ms

Is that what you are experiencing ?

It is up to you to detect press and depress by looking at the payload transitionning from 0 to 1 or 1 to 0.

At least this is what I understand from the code in ButtonHandler()

**Reply by arendst** [reply]
If the button is wrongly configured in the GPIO (button, button_i) it IS always pressed and detected as such.

Depending on how the button is hardwired change the GPIO config accordingly.

**Comment by Heizelmann** [comment]
OK, my understanding of FUNC_BUTTON_PRESSED was wrong. Would be good to explain this in the documentation.

**Reply by Heizelmann** [reply]
By the way, how can I trigger FUNC_BUTTON_MULTI_PRESSED ?

**Reply by barbudor** [reply]
By pressing multiple times on a button ?

```
      case FUNC_BUTTON_MULTI_PRESSED:
        AddLog(LOG_LEVEL_DEBUG, PSTR("TEST: BUTTON_MULTI_PRESSED_%d=%d"), XdrvMailbox.index, XdrvMailbox.payload);
        break;
```

```
19:31:42.506 BTN: Button1 multi-press 1
19:31:43.008 TEST: BUTTON_MULTI_PRESSED_0=1
19:31:43.009 SRC: Button
19:31:43.013 MQT: stat/nodemcu/POWER = {"POWER":"ON"}
19:31:43.015 MQT: stat/nodemcu/POWER = ON (retained)
19:31:44.115 BTN: Button1 multi-press 1
19:31:44.316 BTN: Button1 multi-press 2
19:31:44.820 TEST: BUTTON_MULTI_PRESSED_0=2
19:31:55.154 BTN: Button1 multi-press 1
19:31:55.411 BTN: Button1 multi-press 2
19:31:55.563 BTN: Button1 multi-press 3
19:31:56.075 TEST: BUTTON_MULTI_PRESSED_0=3
19:31:58.908 BTN: Button1 multi-press 1
19:31:59.109 BTN: Button1 multi-press 2
19:31:59.310 BTN: Button1 multi-press 3
19:31:59.512 BTN: Button1 multi-press 4
19:32:00.016 TEST: BUTTON_MULTI_PRESSED_0=4
19:32:00.885 BTN: Button1 multi-press 1
19:32:01.089 BTN: Button1 multi-press 2
19:32:01.293 BTN: Button1 multi-press 3
19:32:01.498 BTN: Button1 multi-press 4
19:32:01.650 BTN: Button1 multi-press 5
19:32:02.153 TEST: BUTTON_MULTI_PRESSED_0=5
19:32:03.402 BTN: Button1 multi-press 1
19:32:03.654 BTN: Button1 multi-press 2
19:32:03.856 BTN: Button1 multi-press 3
19:32:04.212 BTN: Button1 multi-press 4
19:32:04.467 BTN: Button1 multi-press 5
19:32:04.723 BTN: Button1 multi-press 6
19:32:05.235 TEST: BUTTON_MULTI_PRESSED_0=6
```

> 19:31:42.506 BTN: 

logs from the buton driver

> 19:32:05.235 TEST: BUTTON_MULTI_PRESSED

logs from within my test driver from code above

**Reply by Heizelmann** [reply]
For me this does not work. FUNC_BUTTON_PRESSED is OK, but I can not get a FUNC_BUTTON_MULTI_PRESSED. 
My button is normally high and open, and with external pull-up.
Could it be, that this function is disabled in some way. 

Here my code snippet

> 
    case FUNC_BUTTON_MULTI_PRESSED:
       AddLog(LOG_LEVEL_INFO, PSTR("Multi Button pressed:%d"),XdrvMailbox.payload);       
       result = true;
       break;
    case FUNC_BUTTON_PRESSED:
        if (!XdrvMailbox.index && ((PRESSED == XdrvMailbox.payload) && (NOT_PRESSED == Button.last_state[XdrvMailbox.index]))) {
           AddLog(LOG_LEVEL_INFO, PSTR("Button pressed:%d"),XdrvMailbox.payload);
         }
        result = true;
      break;
>
19:46:09.678 Button pressed:0
19:46:09.927 Button pressed:0
19:46:10.177 Button pressed:0
19:46:10.430 Button pressed:0
19:46:10.680 Button pressed:0

**Reply by barbudor** [reply]
Could you have disabled multi-press with [`SetOption13 1`](https://tasmota.github.io/docs/Commands/#setoption13) ?

I took my nodemcu
Configured GPIO0 as Button_1 and only that (no other gpio configuerd)

``` C++
bool Xdrv126(uint32_t function) {
  bool result = false;

  if (FUNC_PRE_INIT == function) {
    DriverTestPreInit();
  } else if (DriverTestReady) {
    switch (function) {
      case FUNC_BUTTON_PRESSED:
        //AddLog(LOG_LEVEL_DEBUG, PSTR("TEST: BUTTON_PRESSED_%d=%d"), XdrvMailbox.index, XdrvMailbox.payload);
        break;
      case FUNC_BUTTON_MULTI_PRESSED:
        AddLog(LOG_LEVEL_DEBUG, PSTR("TEST: BUTTON_MULTI_PRESSED_%d=%d"), XdrvMailbox.index, XdrvMailbox.payload);
        break;
      case FUNC_JSON_APPEND:
        DriverTestShow(1);
        break;
#ifdef USE_WEBSERVER
      case FUNC_WEB_SENSOR:
        DriverTestShow(0);
        break;
#endif  // USE_WEBSERVER
    }
  }
  return result;
}
```

**Reply by Heizelmann** [reply]
Sorry, no success. My Button is on GPIO13, active low and with external pullup resistor. SetOption13 0 has no effect. Config with type button or button i

**Comment by Heizelmann** [comment]
You may look at my complete code under xdrv_61.ino

**Reply by barbudor** [reply]
Got it
The reason is that you have to service either FUNC_BUTTON_PRESSED or FUNC_BUTTON_MULTI_PRESSED
As you have answered true to FUNC_BUTTON_PRESSED, you are not been called for FUNC_BUTTON_MULTI_PRESSED

The believe idea is that
- if your driver doesn't need to manage multi-press, it handles FUNC_BUTTON_PRESSED
- if has to manage multi press, it handles FUNC_BUTTON_MULTI_PRESSED, the payload will always telle how many press there was

But not both

**Reply by Heizelmann** [reply]
FUNC_BUTTON_MULTI_PRESSED is excactly what I am looking for. It also triggers only when button pressed. In my opinion the function FUNC_BUTTON_PRESSED is named wrong with the current implementation, it should be named FUNC_BUTTON_SCAN, FUNC_BUTTON_SAMPLED or something like this.

Thanks a lot!

**Reply by barbudor** [reply]
Probably a legacy issue
created with a name, then multipress was added but renaming the original was out of scope/time/need..
